h2/src/main/org/h2/server/TcpServerThread.java

```diff
@@ -353,10 +354,44 @@ private void process() throws IOException {
             int id = transfer.readInt();
             Command command = (Command) cache.getObject(id, false);
             setParameters(command);
+            boolean supportsGeneratedKeys = clientVersion >= Constants.TCP_PROTOCOL_VERSION_17;
+            boolean writeGeneratedKeys = supportsGeneratedKeys;
+            Object generatedKeysRequest;
+            if (supportsGeneratedKeys) {
+                int type = transfer.readInt();
+                switch (type) {
+                default:
+                    generatedKeysRequest = false;
+                    writeGeneratedKeys = false;
+                    break;
+                case 1:
```

Comment:
"Magic" numbers? Can we have at least constants for them to somewhat link code from here and CommandRemote? Ideally, I would re-factor all this read / write logic (and type casting) into a separate enum, something like this:
`/*
 * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
 * and the EPL 1.0 (http://h2database.com/html/license.html).
 * Initial Developer: H2 Group
 */
package org.h2.result;

import org.h2.value.Transfer;
import java.io.IOException;

/**
 * @author <a href='mailto:andrei.tokar@gmail.com'>Andrei Tokar</a>
 */
public enum GeneratedKeysRequestType {
    NONE {
        @Override
        protected boolean accept(Object generatedKeysRequest) {
            return Boolean.FALSE.equals(generatedKeysRequest);
        }

        @Override
        public Object read(Transfer transfer) {
            return Boolean.FALSE;
        }
    },
    AUTO {
        @Override
        protected boolean accept(Object generatedKeysRequest) {
            return Boolean.TRUE.equals(generatedKeysRequest);
        }

        @Override
        public Object read(Transfer transfer) {
            return Boolean.TRUE;
        }
    },
    INDEXES {
            @Override
            protected boolean accept(Object generatedKeysRequest) {
                return generatedKeysRequest instanceof int[];
            }

            @Override
            public void write(Transfer transfer, Object generatedKeysRequest) throws IOException {
                assert generatedKeysRequest instanceof int[];
                super.write(transfer, generatedKeysRequest);
                int[] keys = (int[]) generatedKeysRequest;
                transfer.writeInt(keys.length);
                for (int key : keys) {
                    transfer.writeInt(key);
                }
            }

            @Override
            public Object read(Transfer transfer) throws IOException {
                int len = transfer.readInt();
                int[] keys = new int[len];
                for (int i = 0; i < len; i++) {
                    keys[i] = transfer.readInt();
                }
                return keys;
            }
    },
    NAMES {
            @Override
            protected boolean accept(Object generatedKeysRequest) {
                return generatedKeysRequest instanceof String[];
            }

            @Override
            public void write(Transfer transfer, Object generatedKeysRequest) throws IOException {
                assert generatedKeysRequest instanceof String[];
                super.write(transfer, generatedKeysRequest);
                String[] keys = (String[]) generatedKeysRequest;
                transfer.writeInt(keys.length);
                for (String key : keys) {
                    transfer.writeString(key);
                }
            }

            @Override
            public Object read(Transfer transfer) throws IOException {
                int len = transfer.readInt();
                String[] keys = new String[len];
                for (int i = 0; i < len; i++) {
                    keys[i] = transfer.readString();
                }
                return keys;
            }
    };

    public static GeneratedKeysRequestType readFrom(Transfer transfer) throws IOException {
        return GeneratedKeysRequestType.values()[transfer.readInt()];
    }

    public static GeneratedKeysRequestType fromObject(Object generatedKeysRequest) {
        for (GeneratedKeysRequestType type : values()) {
            if(type.accept(generatedKeysRequest)) {
                return type;
            }
        }
        return NONE;
    }

    public void write(Transfer transfer, Object generatedKeysRequest) throws IOException {
        transfer.writeInt(ordinal());
    }

    public abstract Object read(Transfer transfer) throws IOException;

    protected abstract boolean accept(Object generatedKeysRequest);
}
`

Reply:
I wrote a new class with these numbers as constants and moved some logic into it.

Enumeration is not suitable here, because there is no efficient way to get enumeration constant by its ordinal value. I can cache array of all values for this, but this cache may be cleared by Tomcat on unloading of web application, this can be also handled, but in the end we got a more dirty code than now.

Reply:
It's not that bad just a cloning of $VALUES array, in our case 16 or 32 bytes + array header, but you have a point.
As far as Tomcat-related remark goes (just my curiosity), will they really clear this?
`private static final GeneratedKeysMode VALUES[] = values();`
On what grounds and at which point in time?

Reply:
`TestClearReferences` detects such non-primitive array as an error. Logic in this test should be copied from a some Tomact version.